polkadot/node/core/approval-voting-parallel/src/lib.rs

```diff
@@ -0,0 +1,950 @@
+		},
+		subsystem.db.clone(),
+		subsystem.keystore.clone(),
+		subsystem.sync_oracle,
+		subsystem.metrics.approval_voting_metrics(),
+		subsystem.spawner.clone(),
+		"approval-voting-parallel-db",
+		"approval-voting-parallel",
+		subsystem.clock.clone(),
+	)
+	.await?;
+
+	Ok((to_approval_voting_worker, to_approval_distribution_workers))
+}
+
+// The main run function of the approval voting subsystem.
+#[overseer::contextbounds(ApprovalVotingParallel, prefix = self::overseer)]
+async fn run<Context>(
+	mut ctx: Context,
+	subsystem: ApprovalVotingParallelSubsystem,
+) -> SubsystemResult<()> {
+	let subsystem_enabled = subsystem.subsystem_enabled;
+	let mut metrics_watcher = MetricsWatcher::new(subsystem.metrics.clone());
+	gum::info!(
+		target: LOG_TARGET,
+		subsystem_enabled,
+		"Starting workers"
+	);
+
+	let (to_approval_voting_worker, to_approval_distribution_workers) =
+		start_workers(&mut ctx, subsystem, &mut metrics_watcher).await?;
+
+	gum::info!(
+		target: LOG_TARGET,
+		subsystem_enabled,
+		"Starting main subsystem loop"
+	);
+
+	// Main loop of the subsystem, it shouldn't include any logic just dispatching of messages to
+	// the workers.
+	run_main_loop(
+		ctx,
+		subsystem_enabled,
+		to_approval_voting_worker,
+		to_approval_distribution_workers,
+		metrics_watcher,
+	)
+	.await
+}
+
+#[overseer::contextbounds(ApprovalVotingParallel, prefix = self::overseer)]
+async fn run_main_loop<Context>(
+	mut ctx: Context,
+	subsystem_enabled: bool,
+	mut to_approval_voting_worker: ToWorker<ApprovalVotingMessage>,
+	mut to_approval_distribution_workers: Vec<ToWorker<ApprovalDistributionMessage>>,
+	metrics_watcher: MetricsWatcher,
+) -> SubsystemResult<()> {
+	loop {
+		futures::select! {
+			next_msg = ctx.recv().fuse() => {
+				let next_msg = match next_msg {
+					Ok(msg) => msg,
+					Err(err) => {
+						gum::info!(target: LOG_TARGET, ?err, "Approval voting parallel subsystem received an error");
+						return Err(err);
+					}
+				};
+				if !subsystem_enabled {
+					gum::trace!(target: LOG_TARGET, ?next_msg, "Parallel processing is not enabled skipping message");
+					continue;
+				}
+
+				match next_msg {
+					FromOrchestra::Signal(msg) => {
+						if matches!(msg, OverseerSignal::ActiveLeaves(_)) {
+							metrics_watcher.collect_metrics();
+						}
+
+						for worker in to_approval_distribution_workers.iter_mut() {
+							worker
+								.send_signal(msg.clone()).await?;
+						}
+
+						to_approval_voting_worker.send_signal(msg.clone()).await?;
+						if matches!(msg, OverseerSignal::Conclude) {
+							break;
+						}
+					},
+					FromOrchestra::Communication { msg } => match msg {
+						// The message the approval voting subsystem would've handled.
+						ApprovalVotingParallelMessage::ApprovedAncestor(_, _,_) |
+						ApprovalVotingParallelMessage::GetApprovalSignaturesForCandidate(_, _)  => {
+							// Safe to unwrap because we know the message is of the right type.
```

Comment:
these three messages should always be convertable into ApprovalVotingMessage if they aren't something crazy happened and we better panic since there is no way forward.